### PR TITLE
Add unit tests for IArrayElementReferenceExpression and make the API …

### LINKED ISSUE #22006: API review for IArrayElementReferenceExpression
https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/Core/Portable/Operations/IArrayElementReferenceExpression.cs

Add unit tests when making the API public

### PR DESCRIPTION
**Pull request by mavasani**
…public again

Fixes #22006

### CODE REVIEW

**Comment by mavasani** [comment]
@dotnet-bot test this please

**Comment by mavasani** [comment]
@dotnet-bot retest this please

**Comment by AlekseyTs** [comment]
>                 Instance Receiver: IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: C) (Syntax: 'a')

Not related to this PR, but how one would distinguish a ```this``` reference from a reference to the object being created? 

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:349 in e47a02a. [](commit_id = e47a02afa6b665dd0d4a53384378f4e1fd63a3ce, deletion_comment = False)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArrayElementReferenceExpression.vb [hunk 1]
```diff
@@ -0,0 +1,773 @@
...
+            VerifyOperationTreeAndDiagnosticsForTest(Of InvocationExpressionSyntax)(source, expectedOperationTree, expectedDiagnostics)
+        End Sub
+
+        <CompilerTrait(CompilerFeature.IOperation)>
+        <Fact, WorkItem(22006, "https://github.com/dotnet/roslyn/issues/22006")>
+        Public Sub ArrayElementReference_ExplicitUserDefinedConversionInArrayReference()
+            Dim source = <![CDATA[
+Class C
+    Public Sub F(o As Object, x As Integer)
+        Dim a = DirectCast(o, String())(x)'BIND:"DirectCast(o, String())(x)"
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArrayElementReferenceExpression.vb]: DirectCast never uses user defined conversion, it looks like scenario differs from the name of the test.

**Comment by AlekseyTs** [comment]
Done with review pass.

**Comment by mavasani** [comment]
>                 Instance Receiver: IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: C) (Syntax: 'a')

Good point, I am not sure such an API exists currently. Filed [https://github.com/dotnet/roslyn/issues/22446](https://github.com/dotnet/roslyn/issues/22446) to discuss possibility of adding one.

---
In reply to: [333237206](https://github.com/dotnet/roslyn/pull/22410#issuecomment-333237206) [](ancestors = 333237206)

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:349 in e47a02a. [](commit_id = e47a02afa6b665dd0d4a53384378f4e1fd63a3ce, deletion_comment = False)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArrayElementReferenceExpression.vb (hunk 1 above)
Comment (mavasani) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArrayElementReferenceExpression.vb]: Yes, it is a mistake while cloning the test, will rename.

---
In reply to: [141971512](https://github.com/dotnet/roslyn/pull/22410#discussion_r141971512) [](ancestors = 141971512)

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 3)

**Comment by mavasani** [comment]
@jcouv @cston - Can I get one more review and ask mode approval please?

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArrayElementReferenceExpression.vb [hunk 2]
```diff
@@ -0,0 +1,773 @@
...
+        Public Sub ArrayElementReference_ExplicitConversionInArrayReference()
+            Dim source = <![CDATA[
+Class C
+    Public Sub F(o As Object, x As Integer)
+        Dim a = DirectCast(o, String())(x)'BIND:"DirectCast(o, String())(x)"
+    End Sub
+
+    Public Shared Widening Operator CType(ByVal c As C) As String()
+        Return Nothing
+    End Operator
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArrayElementReferenceExpression.vb]: Is the `Operator` needed?

**Review by cston — APPROVED** [review]

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArrayElementReferenceExpression.vb (hunk 2 above)
Comment (mavasani) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArrayElementReferenceExpression.vb]: No, will remove it.

**Comment by mavasani** [comment]
@jcouv - approved to merge for 15.5?

**Comment by mavasani** [comment]
Tagging @jaredpar for ask mode approval

**Comment by mavasani** [comment]
Ping @jcouv @jaredpar for ask mode approval.

**Comment by jaredpar** [comment]
approved

**Pull request merged**